src/rpc/methods/eth/filter/mod.rs

```diff
@@ -421,6 +421,78 @@ impl EthEventHandler {
 
         Ok(collected_events)
     }
+
+    pub async fn eth_get_events_for_event_filter<DB: Blockstore + Send + Sync + 'static>(
+        &self,
+        ctx: &Ctx<DB>,
+        filter: &EventFilter,
+        skip_event: SkipEvent,
+    ) -> anyhow::Result<Vec<CollectedEvent>> {
+        let mut collected_events = vec![];
+        match &filter.tipsets {
+            ParsedFilterTipsets::Hash(block_hash) => {
+                let tipset = get_tipset_from_hash(ctx.chain_store(), block_hash)?;
+                let tipset = Arc::new(tipset);
+                Self::collect_events(
+                    ctx,
+                    &tipset,
+                    Some(filter),
+                    skip_event,
+                    &mut collected_events,
+                )
+                .await?;
+            }
+            ParsedFilterTipsets::Key(tsk) => {
+                let tipset = Arc::new(Tipset::load_required(ctx.store(), tsk)?);
+                Self::collect_events(
+                    ctx,
+                    &tipset,
+                    Some(filter),
+                    skip_event,
+                    &mut collected_events,
+                )
+                .await?;
+            }
+            ParsedFilterTipsets::Range(range) => {
+                let max_height = if *range.end() == -1 {
+                    // heaviest tipset doesn't have events because its messages haven't been executed yet
+                    ctx.chain_store().heaviest_tipset().epoch() - 1
+                } else if *range.end() < 0 {
+                    bail!("max_height requested is less than 0")
+                } else if *range.end() > ctx.chain_store().heaviest_tipset().epoch() - 1 {
+                    // we can't return events for the heaviest tipset as the transactions in that tipset will be executed
+                    // in the next non-null tipset (because of Filecoin's "deferred execution" model)
+                    bail!("max_height requested is greater than the heaviest tipset");
```

Comment:
It'd be easier to understand if those two would be under `ensure!` at the beginning of the match arm (preconditions).